### PR TITLE
Fix lock and semaphore timeouts

### PR DESCRIPTION
**Pull request by marcin-krystianc**
The logic for calculating the `WaitTime` was broken.
I think that this bug was inherited from go library. It is fixed there already (https://github.com/hashicorp/consul/commit/9043966efdab6e8e12c73772de070e25c4a50bc1)

### CODE REVIEW

File: Consul/Lock.cs [hunk 1]
```diff
@@ -251,23 +252,24 @@ public async Task<CancellationToken> Acquire(CancellationToken ct)
 
                     var qOpts = new QueryOptions()
                     {
-                        WaitTime = Opts.LockWaitTime
+                        WaitTime = Opts.LockWaitTime,
                     };
 
                     var attempts = 0;
-                    var start = DateTime.UtcNow;
```
Comment (marcin-krystianc) [inline Consul/Lock.cs:258]: `DateTime` arithmetic is not a good way of calculating elapsed time, `StopWatch` is more accurate.

File: Consul/Lock.cs [hunk 2]
```diff
@@ -251,23 +252,24 @@ public async Task<CancellationToken> Acquire(CancellationToken ct)
 
                     var qOpts = new QueryOptions()
                     {
-                        WaitTime = Opts.LockWaitTime
+                        WaitTime = Opts.LockWaitTime,
```
Comment (jgiannuzzi) [inline Consul/Lock.cs:255]: Please remove the comma at the end for style consistency with the rest of the project

**Review by jgiannuzzi — COMMENTED** [review]
Great catch, @marcin-krystianc!

I just have one tiny nitpick below, but otherwise it LGTM 👍

**Comment by jgiannuzzi** [comment]
A more in depth explanation of @marcin-krystianc's fix can be found in the [original PR for Consul](https://github.com/hashicorp/consul/pull/4346).

File: Consul/Lock.cs (hunk 2 above)
Comment (marcin-krystianc) [reply Consul/Lock.cs]: 👍

File: Consul/Lock.cs [hunk 3]
```diff
@@ -251,23 +252,24 @@ public async Task<CancellationToken> Acquire(CancellationToken ct)
...
-                            var elapsed = DateTime.UtcNow.Subtract(start);
-                            if (elapsed > qOpts.WaitTime)
+                            var elapsed = sw.Elapsed;
+                            if (elapsed > Opts.LockWaitTime)
                             {
                                 DisposeCancellationTokenSource();
                                 throw new LockMaxAttemptsReachedException("LockTryOnce is set and the lock is already held or lock delay is in effect");
                             }
-                            qOpts.WaitTime -= elapsed;
+
```
Comment (jgiannuzzi) [inline Consul/Lock.cs:271]: Can you remove the extra line?

**Review by jgiannuzzi — APPROVED** [review]
👍

**Pull request merged**